### ISSUE TITLE
Kubescape project supply chain security self assesment

### ISSUE THREAD

**Issue by slashben** [kubescape/kubescape#1116]
## Overview
See https://github.com/cncf/tag-security/issues/1025

We are in the process of implementing supply chain security best practices in the Kubescape project. 

The first phase of this process is to create a self-assessment based on https://www.cncf.io/announcements/2021/05/14/cncf-paper-defines-best-practices-for-supply-chain-security/.

The maintainers will create a list of expected changes and assess the complexity of implementing them.

We will use this issue to track this task. 

cc: @dwertent @alegrey91 @mlieberman85

**Comment by alegrey91** [comment]
Hi all,
Here's the first part of the list with the expected changes/implementations.

**[Securing the Source Code]**

**Verification**
* **require signed commit**: we should implement a pipeline to check for signed commits. Users that wants to contribute, needs to configure `git` with the following command: `git config commit.gpgsign true`. Difficulty: **easy**.
* **enforce full attestation and verification for protected branches**: we should implement a pipeline to verify signatures before merge the PRs. The pipeline needs to perform the following command: `git merge -S --veriy-signatures`. Difficulty: **easy**.

**Automation**
* **prevent committing secrets to the source code repository**: we should implement a pipeline with the `trufflehog` utility. Difficulty: **easy**.
* **define individuals/teams responsible for code and coding conventions**: we should provide `codeowners`, `.gitignore`, `.gitattributes` and `denylist` files into the repositories. Additionally we could configure **pre-commit hooks** that runs **linting** pipelines to check code conventions. Difficulty: **medium**.
* **automate software security scanning and testing**: run unit, functional and e2e tests (already implemented) + SAST & DAST scans. For SAST scans we can use `gosec` & `golangci-lint` (already implemented). Additionally, metadata from SAST scans, should be recorded and linked as hash to create chain of custody. Finally, code coverage and tests results should be published as badges. Difficulty: **medium**.

**Controlled Environments**
* **establish and adhere contribution policies**: we should automate branch deletion from **github** platform. Difficulty: **easy**.
* **define roles aligned to functional responsibilities**: we should create different roles in our organization such as Developers, Owners, Reviewers, Approvers, and Guests. Difficulty: **easy/medium**.
* **enforce an indipendent four-eyes principles**: we should define the following policies: Avoid that the author of a PR can also be the approver of that. At least 2 people (one of them with write access to the branch) needs to approve a PR before merging it. Difficulty: **easy/medium**.
* **use branch protection rules**: define a rule to avoid `git push -f`ing on the branches. Difficulty: **easy/medium**.

**Secure Authentication**
* **enforce MFA**: we should enable MFA and make it mandatory in order to access the repositories on **github**. Difficulty: **easy**.
* **use SSH keys to provide developers access to repositories**: correct me if I'm wrong, but this functionality is already the default on **github**, so we should not implement anything here. Difficulty: **not-applicable**.
* **have key rotation policy**: I don't know if **github** allow this, but in case, we should create a key rotation policy on it. Difficulty: **not-applicable**.
* **use short lived/ephimeral credentials**: same as above. Difficulty: **not-applicable**.

We can start reviewing this and understand how to proceed. WDYT?

**Comment by slashben** [comment]
Yes, looks like the triage we needed here. Good job @alegrey91 . We should meet @mlieberman85 later this week and discuss with him

**Comment by XDRAGON2002** [comment]
Hi @slashben @alegrey91!
Is help still needed on this?

**Comment by alegrey91** [comment]
> Hi @slashben @alegrey91! Is help still needed on this?

Yes, it is @XDRAGON2002.
Do you have something already in mind?
Otherwise we can suggest you an issue to take care of.

**Comment by XDRAGON2002** [comment]
Nothing as of yet, feel free to let me know what I can potentially help out with.

**Comment by XDRAGON2002** [comment]
@alegrey91 @slashben also just as a probable addition to the list, maybe https://github.com/kubescape/kubescape/issues/1249 would be something we are also interested in?

**Comment by alegrey91** [comment]
> Nothing as of yet, feel free to let me know what I can potentially help out with.

A good starting point could be this one:
> automate software security scanning and testing: run unit, functional and e2e tests (already implemented) + SAST & DAST scans. For SAST scans we can use gosec & golangci-lint (already implemented). Additionally, metadata from SAST scans, should be recorded and linked as hash to create chain of custody. Finally, code coverage and tests results should be published as badges.

As first task, you could search for more SAST scan tools and then we can decide which one of them integrate in the pipeline.
`gosec` is already a good candidate, but we could add more tools to the list.
Then for each of them we could think to provide a badge on the `README.md` to show the result if provided.

**Comment by XDRAGON2002** [comment]
Sure, I'll take a look! Thanks!

**Comment by XDRAGON2002** [comment]
@alegrey91 I believe a combination of tooling might be better off for us since they could cover each other's weak points in cases providing a better reach for securing the code.

Taking a look at [this list](https://analysis-tools.dev/tag/go) allows us a lot of good tools to use including:
- `go vet` (out of box)
- `golangci-lint` (already implemented)
- `gosec` (great tool)
- `govulncheck` (based off of go vulnerability database, extremely helpful)
- `gokart` (extra security)

There are many more but a combination of the above mentioned ones would be more than enough to ensure confidence in our tool chain.

These could also be included into a `pre-commit-hook` which ties back to this point:
> define individuals/teams responsible for code and coding conventions: we should provide codeowners, .gitignore, .gitattributes and denylist files into the repositories. Additionally we could configure pre-commit hooks that runs linting pipelines to check code conventions. Difficulty: medium.

Thoughts?

**Comment by alegrey91** [comment]
> These could also be included into a `pre-commit-hook` which ties back to this point:
> 
> > define individuals/teams responsible for code and coding conventions: we should provide codeowners, .gitignore, .gitattributes and denylist files into the repositories. Additionally we could configure pre-commit hooks that runs linting pipelines to check code conventions. Difficulty: medium.
> 
> Thoughts?

Personally, I'm a great fan of `pre-commit-hook`s, but we have to admit that is a bit intrusive for people who want to contribute. What do you think about that @slashben?

**Comment by XDRAGON2002** [comment]
Well, in my opinion, while contributing to a security tool I don't think people would be put off by the fact that their contributions are being audited to ensure that the tool continues to remain secure ;)

I also prefer `pre-commit-hooks` because it allows me to ensure my code formatting and enforceable best practices which in most cases are things which would be requested by the project maintainers to be done anyways, so helps me in making all those changes even before I make an "unclean" commit. Saves a bit of time and hassle as a contributor ensuring I don't have to make a new commit just because I forgot to run a check ;)

**Comment by XDRAGON2002** [comment]
Hi @slashben @alegrey91 were there any updates regarding this?

I just wrapped up a bunch of other issues I've been working on so have a bit more free time on my hands now, let me know what I should get busy with ;)
Thanks!

**Comment by slashben** [comment]
Closing this since the original CNCF supply-chain self-assessment exercise effectively wrapped up as Kubescape moved through CNCF Sandbox to Incubating, and the concrete hygiene items (DCO, signed commits, GitGuardian, CodeRabbit, License/Dependency/Security Analysis CI checks, SBOM publishing) are now in place across the org without this single tracking issue driving them. Anyone is welcome to reopen for a fresh audit.
